File: Grammars/Extensions.vb
Imports System.Runtime.CompilerServices
Imports System.ComponentModel
Imports System.Collections.ObjectModel

Friend Module Extensions

    <Extension>
    Public Sub AddRange(Of T)(ByRef extended As BindingList(Of T), ByVal items As IEnumerable(Of T))
        For Each item As T In items
            extended.Add(item)
        Next
    End Sub

    <Extension>
    Public Sub AddRange(Of T)(ByRef extended As ObservableCollection(Of T), ByVal items As IEnumerable(Of T))
        For Each item As T In items
            extended.Add(item)
        Next
    End Sub

    <Extension>
    Public Function ToDictionary(Of K, V)(ByVal extended As NLua.LuaTable) As Dictionary(Of K, V)
        Dim result As New Dictionary(Of K, V)

        For i As Int32 = 0 To extended.Keys.Count - 1
            result.Add(CType(extended.Keys(i), K), CType(extended.Values(i), V))
        Next

        Return result
    End Function

    <Extension>
    Public Function ToList(Of T)(ByVal extended As NLua.LuaTable) As List(Of T)
        Dim result As New List(Of T)

        For i As Int32 = 0 To extended.Values.Count - 1
            result.Add(CType(extended.Values(i), T))
        Next

        Return result
    End Function

End Module

File: Grammars/Extensions.cs
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using NLua;

namespace Grammars
{
    internal static class Extensions
    {
        public static void AddRange<T>(this BindingList<T> extended, IEnumerable<T> items)
        {
            foreach (var item in items)
            {
                extended.Add(item);
            }
        }

        public static void AddRange<T>(this ObservableCollection<T> extended, IEnumerable<T> items)
        {
            foreach (var item in items)
            {
                extended.Add(item);
            }
        }

        public static Dictionary<K, V> ToDictionary<K, V>(this LuaTable extended)
        {
            var result = new Dictionary<K, V>();

            for (var i = 0; i <= extended.Keys.Count - 1; i++)
            {
                result.Add(extended.Keys.Cast<K>().ElementAt(i), extended.Values.Cast<V>().ElementAt(i));
            }

            return result;
        }

        public static List<T> ToList<T>(this LuaTable extended)
        {
            var result = new List<T>();

            for (var i = 0; i <= extended.Values.Count - 1; i++)
            {
                result.Add(extended.Values.Cast<T>().ElementAt(i));
            }

            return result;
        }
    }
}